### PR TITLE
[WIP] Matcher module integration (async module integration)

### PR DESCRIPTION
**Pull request by rexagod**
Continued from #1198 

Getting this one in would make for a nice example of using asynchronous modules inside IS!

cc @jywarren @publiclab/is-reviewers

### CODE REVIEW

**Comment by rexagod** [comment]
https://github.com/publiclab/image-sequencer/blob/7bed44e4c2ed413ddb1fe93f74c9326cc7889d1b/src/modules/Matcher/Module.js#L4-L29

So basically the matcher await is skipped and the later changePixel methods aren't "awaited" for the promise to be resolved. I tried using thenables and asyncs but it seems the PixelManipulation does not support laxy evals as of now? Also note that making pixel manipulation async would in turn require the "draw" function to be made asynchronous too, which would make the `input.PixelManipulation` return a promise instead of an object that'll need to be resolved.

I've been trying to find out some workarounds for this, and was wondering if any of @publiclab/is-reviewers have any ideas regarding the same?

Thanks!

**Comment by rexagod** [comment]
Also, I'm aware of the MinifyImage module using an async fn. However, the point where these two cases differ is that I can't actually include the matcher instantiation **inside** the draw fn (as done below) because (unlike the code below) I only need to instantiate the Matcher object once **outside** the draw fn (evaluating 500 points at every draw won't be feasable at all).

https://github.com/publiclab/image-sequencer/blob/1b26f7165465b68010c885a9f111296838447789/src/modules/MinifyImage/Module.js#L67-L81

**Comment by rexagod** [comment]
Copying in from today's opencall.

cc @jywarren 

***

- wrapup! 
Matcher/IS - Pranshu
https://github.com/publiclab/image-sequencer/pull/1226
https://github.com/publiclab/image-sequencer/blob/7bed44e4c2ed413ddb1fe93f74c9326cc7889d1b/src/modules/Matcher/Module.js#L41
https://github.com/publiclab/image-sequencer/blob/main/src/modules/_nomodule/PixelManipulation.js#L63
https://github.com/publiclab/image-sequencer/blob/main/src/modules/DecodeQr/Module.js - no PixelManipulation
https://github.com/publiclab/image-sequencer/blob/main/CONTRIBUTING.md

```js
var matcher
matcher = new Matcher(
'img.png',
'img2.png',
async function(promise) {
-...await...
-// matcher setup complete
-matcher.doSomething(); // ? OK but also outside
Promise.resolve(promise).then(function done(params) {

var points = ...
this.output = originalImage();
callback();

});
},
query: "corners"
)

var points = ...
this.output = originalImage();
callback();
```

**Comment by rexagod** [comment]
Copying this in from the email sent to @jywarren earlier.

***
> Hello Jeff!

Hope this email finds you well. I actually need to bring to your attention that apparently there aren't any "heavyweight computing" modules that require the execution flow to pause until they are resolved, under the constraint that their respective constructors can only be run *once* (i.e. outside the draw function), Matcher being the first. Therefore, I suppose https://github.com/publiclab/image-sequencer/pull/1226 requires a little more than we expected, i.e., a refactor of the image sequencer to readily support such modules, paving way to a (major?) minor release, a bit of context in the docs, and increase in the overall flexibility of the image-sequencer, not only for synchronous modules, but asynchronous as well!

I believe that this "refactor" isn't something we can avoid, since now or in the future, image-sequencer would demand modules that need more computing power, hence the asynchronity, in order to maintain its versatility.

Also, I can setup a hard coded demo inside the matcher for the tulip image if folks are interested *right now* to see how matcher would work inside image-sequencer, but, obviously, I'm more interested in the dynamic one (a robust approach would be to run matcher at the time the load-image/replace-image module is run, and work with those results)!

Let's make this happen! I'm eager to hear your thoughts on the same!

Thanks!

**Comment by jywarren** [comment]
@HarshKhandeparkar I wonder if this is an interesting project you may want to take a look at... some real big challenges here. We could make a `hall-of-fame` task from it. Just in case you may find it interesting!!!

**Comment by HarshKhandeparkar** [comment]
What is the matcher module?

**Comment by jywarren** [comment]
It compares 2 images and finds common points between them (like 2 photos of
the same object from different angles) then it overlays one on the other to
"stitch" them together. It uses https://github.com/publiclab/matcher-core !!

**Comment by HarshKhandeparkar** [comment]
Then what would be the output?

**Comment by HarshKhandeparkar** [comment]
Oh ok. So like map stitching?

**Comment by HarshKhandeparkar** [comment]
I think this would be a bit hard for GCI. Only 20days to go. Research itself would take me about 3 or 4 days. Then implementing, then bug fixing and ... well, making it work with GIFs(I dug my own grave). If someone else wants to do this for GCI, let them do it. 
If no one does, then maybe I'll do it after GCI.

**Comment by jywarren** [comment]
OK, thanks in any case! :-)

File: examples/index.html [hunk 1]
```diff
@@ -28,6 +28,10 @@
   <script src="../dist/image-sequencer.js" charset="utf-8"></script>
   <script src="../dist/image-sequencer-ui.js" charset="utf-8"></script>
   <script src="../node_modules/bootstrap-colorpicker/dist/js/bootstrap-colorpicker.js"></script>
+
+  <!-- for matcher module -->
+  <script src="orb.core.min.js"></script>
```
Comment (HarshKhandeparkar) [inline examples/index.html:33]: What is this and why does it have to be in the UI part?

**Comment by HarshKhandeparkar** [comment]
Converted to draft for now :)

**Pull request closed without merging**